File: RPI/embeded_final_project/capture_image/capture_image.py
import requests
import cv2
import numpy as np
import os
from datetime import datetime
import shutil
# ESP32相機的網址
# esp32_camera_url = 'http://192.168.0.21/capture' # home
# esp32_camera_url = 'http://192.168.50.117/capture' # lab
# esp32cam_url = "http://172.20.10.2/capture"     # mobile
esp32cam_url = "http://192.168.223.223/capture" # 士倫

# 設置圖片保存的資料夾
output_folder = "/home/user/embeded_final_project/capture_image/output_capture_images/"
# 先清空整個資料夾
shutil.rmtree(output_folder, ignore_errors=True)
if not os.path.exists(output_folder):
    os.makedirs(output_folder)

# 預先分配一個大緩衝區，用於存儲接收到的影像數據
max_image_size = 1024 * 1024  # 假設最大影像大小為1MB
buffer = bytearray(max_image_size)

count = 0

print("start capture")
def capture_image():
    global count  # 使用 global 關鍵字宣告 count 為全域變數

    with requests.Session() as session:  # 使用 requests.Session 建立連線池
        while (count <= 2):  # 使用 每次執行拍三張(count=0, 1, 2) 來持續執行迴圈
            count += 1
            print("\nloop: {}".format(count))

            try:
                # 發送GET請求獲取影像數據
                response = session.get(esp32_camera_url)
                if response.headers['Content-Type'] == 'image/jpeg':
                    print("response.headers OK!")

                    # 檢查請求是否成功
                    if response.status_code == 200:
                        print("response.status OK!")
                        # 將返回的影像數據存儲到預先分配的緩衝區
                        image_data = response.content
                        if len(image_data) <= max_image_size:
                            buffer[:len(image_data)] = image_data
                            nparr = np.frombuffer(buffer[:len(image_data)], np.uint8)
                            image = cv2.imdecode(nparr, cv2.IMREAD_COLOR)

                            # 檢查影像尺寸是否有效
                            if image is not None and image.shape[0] > 0 and image.shape[1] > 0:
                                print("image.shape OK!")

                                # 顯示影像
                                cv2.imshow('ESP32 Camera Image', image)

                                # 保存影像到指定資料夾
                                timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
                                image_filename = os.path.join(output_folder, f"capture_{timestamp}_{count}.jpg")
                                cv2.imwrite(image_filename, image)
                                print(f"Image saved as {image_filename}")

                                cv2.waitKey(100)  # 等待0.1秒再顯示下一張圖像
                            else:
                                print("Invalid image data received")
                        else:
                            print("Image size exceeds buffer size")
                    else:
                        print(f"Failed to fetch image. Status code: {response.status_code}")
                else:
                    print(f"Unexpected content type: {response.headers['Content-Type']}")

            except requests.exceptions.ConnectionError as e:
                print(f"Connection error: {e}")
            except Exception as e:
                print(f"Error: {e}")

if __name__ == "__main__":
    capture_image()

File: RPI/embeded_final_project/capture_image/capture_image.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <curl/curl.h>
#include <opencv2/opencv.hpp>
#include <sys/stat.h>
#include <sys/types.h>
#include <time.h>
#include <dirent.h>
#include <unistd.h>

using namespace cv;

#define MAX_IMAGE_SIZE 1024 * 1024 // 假設最大影像大小為1MB
#define OUTPUT_FOLDER "/home/user/embeded_final_project/capture_image/output_capture_images/"
#define esp32cam_url "http://192.168.50.117/capture" // 根據esp32cam的網址調整

struct MemoryStruct
{
    unsigned char *memory;
    size_t size;
};

static size_t WriteMemoryCallback(void *contents, size_t size, size_t nmemb, void *userp)
{
    size_t realsize = size * nmemb;
    struct MemoryStruct *mem = (struct MemoryStruct *)userp;

    if (mem->size + realsize > MAX_IMAGE_SIZE)
    {
        realsize = MAX_IMAGE_SIZE - mem->size;
    }

    memcpy(&(mem->memory[mem->size]), contents, realsize);
    mem->size += realsize;

    return realsize;
}

void fetch_and_process_image(CURL *curl, int count)
{
    CURLcode res;
    struct MemoryStruct chunk;

    chunk.memory = (unsigned char *)malloc(MAX_IMAGE_SIZE);
    chunk.size = 0;

    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteMemoryCallback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *)&chunk);

    res = curl_easy_perform(curl);

    if (res != CURLE_OK)
    {
        fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
    }
    else
    {
        printf("response.headers OK!\n");

        // 檢查影像尺寸是否有效
        if (chunk.size > 0)
        {
            printf("response.status OK!\n");

            std::vector<uchar> data(chunk.memory, chunk.memory + chunk.size);
            Mat image = imdecode(data, IMREAD_COLOR);

            if (!image.empty() && image.cols > 0 && image.rows > 0)
            {
                printf("image.shape OK!\n");

                // 顯示影像 // 顯不出來
                // imshow("ESP32 Camera Image", image);

                // 保存影像到指定資料夾
                time_t now = time(NULL);
                struct tm *t = localtime(&now);
                char timestamp[32];
                strftime(timestamp, sizeof(timestamp), "%Y%m%d_%H%M%S", t);

                char image_filename[128];
                snprintf(image_filename, sizeof(image_filename), OUTPUT_FOLDER "/capture_%s_%d.jpg", timestamp, count);
                imwrite(image_filename, image);
                printf("Image saved as %s\n\n", image_filename);

                waitKey(100); // 等待0.1秒再顯示下一張圖像
            }
            else
            {
                printf("Invalid image data received\n");
            }
        }
        else
        {
            printf("Image size exceeds buffer size\n");
        }
    }

    free(chunk.memory);
}

void clear_output_folder()
{
    struct dirent *entry;
    DIR *dir = opendir(OUTPUT_FOLDER);

    if (dir == NULL)
    {
        return;
    }

    while ((entry = readdir(dir)) != NULL)
    {
        char filepath[256];
        snprintf(filepath, sizeof(filepath), OUTPUT_FOLDER "/%s", entry->d_name);

        // 忽略 "." 和 ".."
        if (strcmp(entry->d_name, ".") != 0 && strcmp(entry->d_name, "..") != 0)
        {
            remove(filepath);
        }
    }
    closedir(dir);
}

int main(void)
{
    CURL *curl;
    CURLcode res;

    // 創建輸出資料夾
    struct stat st = {0};
    if (stat(OUTPUT_FOLDER, &st) == -1)
    {
        mkdir(OUTPUT_FOLDER, 0700);
    }

    // 清空輸出資料夾
    clear_output_folder();

    // 網路初始化
    printf("init network\n");
    curl_global_init(CURL_GLOBAL_DEFAULT);
    curl = curl_easy_init();

    if (curl)
    {
        // 連接網路
        printf("connecting esp32cam network\n\n");
        curl_easy_setopt(curl, CURLOPT_URL, esp32cam_url);
        curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);

        int count = 0;
        while (count < 3) // 設定要連拍幾次
        {
            count++;
            fetch_and_process_image(curl, count);
        }

        curl_easy_cleanup(curl);
    }

    curl_global_cleanup();

    return 0;
}
